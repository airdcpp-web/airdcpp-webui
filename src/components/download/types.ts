import * as UI from 'types/ui';


export type BrowseHandler = (() => void) | undefined;

export type PathDownloadHandler = (targetPath: string, targetFilename?: string) => Promise<any>;

export type LayoutProps = React.PropsWithChildren<{
  menuItems: React.ReactNode[];
  title: string;
  handleBrowse: BrowseHandler;
  t: UI.TranslateF;
}>;
