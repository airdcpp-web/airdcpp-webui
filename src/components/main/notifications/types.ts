export type NotificationLevel = 'error' | 'warning' | 'info' | 'success';
