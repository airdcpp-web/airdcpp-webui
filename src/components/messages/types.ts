import { IntersectionObserverProps } from 'react-intersection-observer';

import * as UI from '@/types/ui';

export interface CommonMessageProps extends Omit<IntersectionObserverProps, 'children'> {
  highlightMenuProps: UI.MessageActionMenuData;
}
