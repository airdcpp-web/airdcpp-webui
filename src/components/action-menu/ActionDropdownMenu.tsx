import * as React from 'react';

import SectionedDropdown, {
  SectionedDropdownProps,
} from '@/components/semantic/SectionedDropdown';
import MenuSection from '@/components/semantic/MenuSection';
import ActionMenuDecorator, {
  ActionMenuDecoratorChildProps,
} from '@/components/action-menu/decorators/ActionMenuDecorator';
import { buildMenu } from './builder/expandMenuBuilder';

export interface ActionMenuProps extends SectionedDropdownProps {
  header?: React.ReactNode;
}

const ActionDropdownMenu = ({
  header,
  children,
  ...other
}: ActionMenuProps & ActionMenuDecoratorChildProps) => (
  <SectionedDropdown {...other}>
    <MenuSection caption={header}>{children()}</MenuSection>
  </SectionedDropdown>
);

const ActionMenuDecorated = ActionMenuDecorator<ActionMenuProps, any, any>(
  ActionDropdownMenu,
  buildMenu,
);

export { ActionMenuDecorated as ActionMenu };
