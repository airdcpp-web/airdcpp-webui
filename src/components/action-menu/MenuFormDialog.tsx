import { useRef } from 'react';
import * as React from 'react';

import Modal, { ModalProps } from '@/components/semantic/Modal';
import Form, { FormSaveHandler } from '@/components/form/Form';

import * as UI from '@/types/ui';

import { IconType } from '@/components/semantic/Icon';
import { useTranslation } from 'react-i18next';
import { translate } from '@/utils/TranslationUtils';

export interface MenuFormDialogProps extends ModalProps {
  title: string;
  icon: IconType;
  fieldDefinitions: UI.FormFieldDefinition[];
  onSave: FormSaveHandler<UI.FormValueMap>;
}

export const MenuFormDialog: React.FC<MenuFormDialogProps> = ({
  fieldDefinitions,
  onSave,
  title,
  icon,
  ...other
}) => {
  const formRef = useRef<Form | null>(null);
  const { t } = useTranslation();
  return (
    <Modal
      {...other}
      title={title}
      onApprove={() => formRef.current!.save()}
      approveCaption={translate('Continue', t, UI.Modules.COMMON)}
      closable={false}
      icon={icon}
      dynamicHeight={true}
    >
      <Form ref={formRef} onSave={onSave} fieldDefinitions={fieldDefinitions} />
    </Modal>
  );
};
