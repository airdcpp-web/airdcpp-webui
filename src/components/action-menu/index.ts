export * from './ActionDropdownMenu';
export * from './DropdownMenu';
export * from '../icon/MenuIcon';
