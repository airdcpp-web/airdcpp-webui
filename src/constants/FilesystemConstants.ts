const MODULE_URL = 'filesystem';
export default {
  MODULE_URL: MODULE_URL,
  LIST_URL: MODULE_URL + '/list_items',
  DIRECTORY_URL: MODULE_URL + '/directory',
  DISK_INFO_URL: MODULE_URL + '/disk_info',

  LOCATION_DOWNLOAD: 'download',
  LOCATION_GENERIC: 'settings',
};
