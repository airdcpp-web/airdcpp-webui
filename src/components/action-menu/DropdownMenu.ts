import TableDropdown, { TableDropdownProps } from 'components/semantic/TableDropdown';

import ActionMenuDecorator, {
  ActionMenuDecoratorProps,
} from './decorators/ActionMenuDecorator';
import DownloadMenuDecorator, {
  DownloadMenuDecoratorProps,
} from './decorators/DownloadMenuDecorator';
import UserMenuDecorator, {
  UserMenuDecoratorProps,
} from './decorators/UserMenuDecorator';

import { DropdownProps } from 'components/semantic/Dropdown';

import * as UI from 'types/ui';
import { ActionMenuProps, ActionMenu } from './ActionDropdownMenu';
import { buildMenu } from './builder/slidingMenuBuilder';

type TableActionMenuDropdownProps = Omit<TableDropdownProps, 'children' | 'caption'>;

export type TableActionMenuProps<ItemDataT extends UI.ActionMenuItemDataValueType> =
  TableActionMenuDropdownProps & ActionMenuDecoratorProps<ItemDataT>;
export const TableActionMenu = ActionMenuDecorator<TableActionMenuDropdownProps, any>(
  TableDropdown,
  buildMenu,
);

export type UserMenuProps = UserMenuDecoratorProps & ActionMenuProps;
export const UserMenu = UserMenuDecorator<DropdownProps>(ActionMenu);

export type TableUserMenuProps = UserMenuDecoratorProps & TableActionMenuDropdownProps;
export const TableUserMenu =
  UserMenuDecorator<TableActionMenuDropdownProps>(TableActionMenu);

export type DownloadMenuProps<ItemDataT extends UI.DownloadableItemInfo> =
  DownloadMenuDecoratorProps<ItemDataT> & ActionMenuProps;

export const DownloadMenu = DownloadMenuDecorator<DropdownProps, any>(ActionMenu);

export type TableDownloadMenuProps<ItemDataT extends UI.DownloadableItemInfo> =
  DownloadMenuDecoratorProps<ItemDataT> & TableActionMenuDropdownProps;

export const TableDownloadMenu = DownloadMenuDecorator<TableActionMenuDropdownProps, any>(
  TableActionMenu,
);
