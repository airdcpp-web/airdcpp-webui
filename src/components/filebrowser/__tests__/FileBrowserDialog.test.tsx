import {
  getConnectedSocket,
  getMockServer,
} from 'airdcpp-apisocket/tests/mock-server.js';

import { jest } from '@jest/globals';
import { renderNode } from 'tests/test-containers';

import * as API from 'types/api';

import { fireEvent, waitFor } from '@testing-library/dom';
import FilesystemConstants from 'constants/FilesystemConstants';
import { FilesystemListContentResponse } from 'tests/mocks/api/filesystem';
import { MenuFormDialog } from 'components/action-menu/MenuFormDialog';
import { createTestModalController, useModalButton } from 'tests/test-dialog-helpers';
import { clickButton, waitForData } from 'tests/test-helpers';
import { setInputFieldValues, setupUserEvent } from 'tests/test-form-helpers';

// tslint:disable:no-empty
describe('FileBrowserDialog', () => {
  let server: ReturnType<typeof getMockServer>;
  const getSocket = async () => {
    const { socket } = await getConnectedSocket(server);

    // Browse dialog
    server.addRequestHandler(
      'POST',
      FilesystemConstants.LIST_URL,
      FilesystemListContentResponse,
    );

    server.addRequestHandler('POST', FilesystemConstants.DIRECTORY_URL, undefined);

    return { socket, server };
  };

  const renderDialog = async (fieldType: API.SettingTypeEnum, defaultValue = '') => {
    const { socket, server, ...other } = await getSocket();

    const onSave = jest.fn(() => Promise.resolve());
    const caption = 'Test dialog';

    const ShareDirectoryDialogTest = () => {
      const FormDefinitions: API.SettingDefinition[] = [
        {
          key: fieldType,
          type: fieldType,
          default_value: defaultValue,
          title: fieldType,
        },
      ];

      const { open, onClose, button } = useModalButton();
      return (
        <>
          {button}
          {open && (
            <MenuFormDialog
              title={caption}
              onSave={onSave}
              icon="test"
              fieldDefinitions={FormDefinitions}
              onClose={onClose}
            />
          )}
        </>
      );
    };

    const renderData = renderNode(<ShareDirectoryDialogTest />, socket);

    const modalController = createTestModalController(renderData);
    return { modalController, onSave, caption, ...renderData, ...other };
  };

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
  });

  describe('directory selection', () => {
    test('should select initial', async () => {
      const { getByText, modalController, queryByText, getByRole, onSave, caption } =
        await renderDialog(API.SettingTypeEnum.DIRECTORY_PATH);

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      expect(fireEvent.click(getByText('Browse'))).toBeTruthy();
      await waitForData('Loading items', queryByText);

      clickButton('Select', getByRole);

      // Save
      await modalController.closeDialogButton('Continue');

      expect(onSave).toHaveBeenCalledTimes(1);
      expect(onSave.mock.calls[0]).toMatchSnapshot();
    }, 100000);

    test('should select child', async () => {
      const { getByText, modalController, queryByText, getByRole, onSave, caption } =
        await renderDialog(API.SettingTypeEnum.DIRECTORY_PATH);

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      expect(fireEvent.click(getByText('Browse'))).toBeTruthy();
      await waitForData('Loading items', queryByText);

      expect(fireEvent.click(getByText('2 folders'))).toBeTruthy();

      clickButton('Select', getByRole);

      // Save
      await modalController.closeDialogButton('Continue');

      expect(onSave).toHaveBeenCalledTimes(1);
      expect(onSave.mock.calls[0]).toMatchSnapshot();
    }, 100000);

    test('should create new directory', async () => {
      const userEvent = setupUserEvent();
      const {
        getByText,
        modalController,
        queryByText,
        getByRole,
        getByLabelText,
        onSave,
        caption,
        container,
      } = await renderDialog(API.SettingTypeEnum.DIRECTORY_PATH);

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      expect(fireEvent.click(getByText('Browse'))).toBeTruthy();
      await waitForData('Loading items', queryByText);

      // Create new directory
      expect(fireEvent.click(getByText('Create directory'))).toBeTruthy();
      await waitFor(() =>
        expect(container.querySelector('.ui.action.input.visible')).toBeTruthy(),
      );

      await setInputFieldValues(
        { userEvent, getByLabelText },
        { 'Create directory': 'New directory' },
      );

      expect(fireEvent.click(getByText('Create'))).toBeTruthy();
      await waitFor(() =>
        expect(container.querySelector('.ui.action.input.visible')).toBeFalsy(),
      );

      // Select the new directory
      clickButton('Select', getByRole);

      // Save
      await modalController.closeDialogButton('Continue');

      expect(onSave).toHaveBeenCalledTimes(1);
      expect(onSave.mock.calls[0]).toMatchSnapshot();
    }, 100000);
  });

  describe('file selection', () => {
    test.skip('should select existing', async () => {
      const { getByText, modalController, queryByText, onSave, caption } =
        await renderDialog(API.SettingTypeEnum.EXISTING_FILE_PATH);

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      expect(fireEvent.click(getByText('Browse'))).toBeTruthy();
      await waitForData('Loading items', queryByText);

      // Go to child directory
      expect(fireEvent.click(getByText('2 folders'))).toBeTruthy();
      await waitForData('Loading items', queryByText);

      // Save
      await modalController.closeDialogButton('empty.txt');

      expect(onSave).toHaveBeenCalledTimes(1);
      expect(onSave.mock.calls[0]).toMatchSnapshot();
    }, 100000);
  });
});
