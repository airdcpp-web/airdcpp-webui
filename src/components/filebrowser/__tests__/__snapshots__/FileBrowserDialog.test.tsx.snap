// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileBrowserDialog directory selection should create new directory 1`] = `
[
  {
    "directory_path": "/New directory/",
  },
  {
    "directory_path": "/New directory/",
  },
]
`;

exports[`FileBrowserDialog directory selection should select child 1`] = `
[
  {
    "directory_path": "/2 folders/",
  },
  {
    "directory_path": "/2 folders/",
  },
]
`;

exports[`FileBrowserDialog directory selection should select initial 1`] = `
[
  {
    "directory_path": "/",
  },
  {
    "directory_path": "/",
  },
]
`;
